### PR TITLE
[mono] Set thread suspend default in mono.proj; default to hybrid

### PR DESCRIPTION
**Pull request by lambdageek**
Instead of detecting in cmake, make a decision in mono.proj

### CODE REVIEW

**Comment by lambdageek** [comment]
This is, in part, motivated by #44806 - MBR needs hybrid (or full coop) suspend in order to be able to inspect interpreter frames.  So we will need hybrid suspend on the mobile platforms.

Previously hybrid suspend was already well tested on Linux and OSX amd64.  And apparently in dotnet/runtime we already used it on Android, too.

If it is not feasible to identify and fix all hybrid suspend issues, we could set the default back to preemptive on affected platforms.

(For MBR we would then launch hot reload sessions with hybrid suspend set via environment variables).

This also sets "hybrid" as the suspend mechanism on wasm, although in single threaded operation it doesn't make a difference.

File: src/mono/CMakeLists.txt [hunk 1]
```diff
@@ -543,6 +539,9 @@ if (GC_SUSPEND STREQUAL "coop")
   set(ENABLE_COOP_SUSPEND 1)
 elseif(GC_SUSPEND STREQUAL "hybrid")
   set(ENABLE_HYBRID_SUSPEND 1)
+elseif(GC_SUSPEND STREQUAL "preemptive")
+else()
+  message(FATAL_ERROR "GC_SUSPEND (set to '${GC_SUSPEND}') must be one of coop, hybrid or preemptive")
```
Comment (CoffeeFlux) [inline src/mono/CMakeLists.txt:544]: If you're removing the option of a default in the main CMake file, it should be removed from [here](https://github.com/dotnet/runtime/blob/master/src/mono/cmake/options.cmake#L70) as well.

**Review by CoffeeFlux — APPROVED** [review]
Hope you're ready to start putting out fires when the CI failures roll in :smile:

File: src/mono/CMakeLists.txt (hunk 1 above)
Comment (lambdageek) [reply src/mono/CMakeLists.txt:544]: @CoffeeFlux I kind of like that it's set in options.cmake to an invalid value.  That way it's obvious if mono.proj is missing a case.  But I guess we could change the string to anything - maybe `"change in mono.proj"`.

Comment (vargaz) [reply src/mono/CMakeLists.txt:544]: I think there should still be a reasonable default so this can be used without using mono.proj.

**Comment by lambdageek** [comment]
wasm failure looks related.  android and windows x64 also possibly related.

File: src/mono/CMakeLists.txt (hunk 1 above)
Comment (lambdageek) [reply src/mono/CMakeLists.txt:554]: ok, added back some code to pick a default (hybrid, unless you're on windows or watch) if GC_SUSPEND isn't set from the outside

File: src/mono/CMakeLists.txt [hunk 2]
```diff
@@ -540,6 +540,16 @@ if (GC_SUSPEND STREQUAL "coop")
 elseif(GC_SUSPEND STREQUAL "hybrid")
   set(ENABLE_HYBRID_SUSPEND 1)
 elseif(GC_SUSPEND STREQUAL "preemptive")
+elseif(GC_SUSPEND STREQUAL "default")
+  # set some kind of fallback default
+  message(WARNING "GC_SUSPEND set to 'default', using a fallback default")
+  if(TARGET_SYSTEM_NAME STREQUAL "watchOS")
```
Comment (vargaz) [inline src/mono/CMakeLists.txt:546]: Not sure this needs a warning.

**Comment by lambdageek** [comment]
Defaulting wasm to "preemptive".  Otherwise it was asserting when trying to insert safepoints in AOT code.  Wasm isn't really preemptive but we don't need safepoints in single-threaded.  (and this is fine for hot reload, too, because the LMF has all the managed interpreted frames that we care about - we know the thread is calling the hot reload icall)

**Comment by lambdageek** [comment]
okay, let's do this, I guess. Looking forward to all the new crash reports over the weekend.

**Comment by marek-safar** [comment]
> Otherwise it was asserting when trying to insert safepoints in AOT code.

Could you create a tracking issue for that if there is not one already?

**Comment by lambdageek** [comment]
Ok, opened https://github.com/dotnet/runtime/issues/47080

**Pull request merged**